src/NUnitFramework/framework/Assert.cs

```diff
@@ -404,7 +404,29 @@ private static void IssueWarning(string message)
         private static readonly StackFilter SystemEnvironmentFilter = new StackFilter(@" System\.Environment\.");
 
         private static string GetStackTrace() =>
-            StackFilter.DefaultFilter.Filter(SystemEnvironmentFilter.Filter(Environment.StackTrace));
+            StackFilter.DefaultFilter.Filter(SystemEnvironmentFilter.Filter(GetEnvironmentStackTraceWithoutThrowing()));
+
+        /// <summary>
+        /// If <see cref="Exception.StackTrace"/> throws, returns "SomeException was thrown by the
+        /// Environment.StackTrace property." See also <see cref="ExceptionExtensions.GetStackTraceWithoutThrowing"/>.
+        /// </summary>
+#if !(NET35 || NETSTANDARD1_4)
+        // https://github.com/dotnet/coreclr/issues/19698 is also currently present in .NET Framework 4.7 and 4.8. A
+        // race condition between threads reading the same PDB file to obtain file and line info for a stack trace
+        // results in AccessViolationException when the stack trace is accessed even indirectly e.g. Exception.ToString.
+        [System.Runtime.ExceptionServices.HandleProcessCorruptedStateExceptions]
+#endif
+        private static string GetEnvironmentStackTraceWithoutThrowing()
+        {
+            try
+            {
+                return Environment.StackTrace;
+            }
+            catch (Exception ex)
+            {
+                return ex.GetType().Name + " was thrown by the Environment.StackTrace property.";
```

Comment:
I thought about it. I'm sure this is safe from user code throwing an exception since the C# rules bind this call to `System.Object.GetType()` which is sealed.

I can't think of any normal exceptions the BCL itself could throw when GetType is called. A type load exception would have prevented the exception from even being instantiated. That leaves exceptions where the BCL is seriously broken, such as the race condition when creating stack traces that hasn't been patched yet in .NET 4.7.2 and 4.8. Those we should only plan for when we have a specific case because there are just too many ways things could go wrong in theory all over the place if the BCL is unreliable.